### PR TITLE
BUG: don't create array with invalid memory in where

### PR DESCRIPTION
**Pull request by juliantaylor**
When creating the tuple view of np.where, make sure the arrays point to
valid memory when they are empty.
Numpy assumes that even empty arrays have valid memory, e.g. in the
source stride == 0 assignments.
Closes gh-8922.

### CODE REVIEW

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -6619,6 +6619,14 @@ def test_string(self):
         assert_equal(np.where(True, a, b), "abcd")
         assert_equal(np.where(False, b, a), "abcd")
 
+    def test_empty_result(self):
+        # pass empty where result through an assignment which reads the data of
+        # empty arrays, error detectable with valgrind, see gh-8922
```
Comment (eric-wieser) [inline numpy/core/tests/test_multiarray.py]: Isn't the real bug here that data is being read from an empty array?

**Comment by juliantaylor** [comment]
We could also put a `N > 0` guard onto https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/lowlevel_strided_loops.c.src#L219 which might be the only place we actually are trying to read from empty arrays.
But this would prevent is from finding these types of bugs.

File: numpy/core/tests/test_multiarray.py (hunk 1 above)
Comment (juliantaylor) [inline numpy/core/tests/test_multiarray.py]: arguably yes, but doing so helps finding these bugs

**Comment by eric-wieser** [comment]
It seems to be that it should be perfectly legal to invoke `PyArray_New` with an empty shape and an invalid data pointer.

Given that it's public API, this patch doesn't really do anything other that hide the issue in the first place. Third-party code could continue to call the constructor in this way.

**Comment by eric-wieser** [comment]
IMO, the bug is either that `@prefix@_@oper@_size@elsize@_srcstride0` is breaking contract when passed `N == 0`, or possibly that `@prefix@_@oper@_size@elsize@_srcstride0` should not even have been called when `N == 0`

**Comment by juliantaylor** [comment]
hm maybe it is better to just fix the function we know does it.

**Comment by eric-wieser** [comment]
Presumably, something like `np.ndarray('', strides=(1000, 1000), shape=(0,0))` could also trigger a similar error?

**Comment by juliantaylor** [comment]
No as numpy gives empty arrays valid (uninitialized) memory for one block regardless. Though this should also trigger in that case harmless valgrind warnings.

File: numpy/core/src/multiarray/lowlevel_strided_loops.c.src [hunk 2]
```diff
@@ -206,6 +206,9 @@ static NPY_GCC_OPT_3 void
 #else
     npy_uint64 temp0, temp1;
 #endif
+    if (N == 0) {
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/lowlevel_strided_loops.c.src]: Could probably do with a comment along the lines of "dereferencing `src` when `N == 0` is not allowed

**Pull request merged**